omegaconf/base.py

```diff
@@ -413,15 +416,71 @@ def _resolve_interpolation_from_parse_tree(
-            )
+        # If the output is not a Node already (e.g., because it is the output of a
+        # custom resolver), then we will need to wrap it within a Node.
+        must_wrap = not isinstance(resolved, Node)
+
+        if isinstance(value, AnyNode):
+            # If the node is untyped, we still need to validate against `None`
+            # in case the `optional` flag is set to `False`.
+            if not value._is_optional() and OmegaConf.is_none(resolved):
+                if throw_on_resolution_failure:
+                    self._format_and_raise(
+                        key=key,
+                        value=resolved,
+                        cause=InterpolationValidationError(
+                            "Non optional field cannot be assigned None"
+                        ),
+                    )
+                return None
+
+        elif isinstance(value, ValueNode):
+            # If the node is typed, validate (and possibly convert) the result.
+            res_value = _get_value(resolved)
+            try:
+                conv_value = value.validate_and_convert(res_value)
+            except ValidationError as e:
+                if throw_on_resolution_failure:
+                    self._format_and_raise(
+                        key=key,
+                        value=res_value,
+                        cause=e,
+                        type_override=InterpolationValidationError,
+                    )
+                return None
+
+            # If the converted value is of the same type, it means that no conversion
+            # was actually needed. As a result, we can keep the original `resolved`
+            # (and otherwise, the converted value must be wrapped into a new node).
+            if type(conv_value) != type(res_value):
+                must_wrap = True
+                resolved = conv_value
+
+        if must_wrap:
+            assert parent is None or isinstance(parent, BaseContainer)
+            try:
+                wrapped = _node_wrap(
+                    type_=value._metadata.ref_type,
+                    parent=parent,
+                    is_optional=value._metadata.optional,
+                    value=resolved,
+                    key=key,
+                    ref_type=value._metadata.ref_type,
+                )
+            except (KeyValidationError, ValidationError) as e:
+                if throw_on_resolution_failure:
+                    self._format_and_raise(
+                        key=key,
+                        value=resolved,
+                        cause=e,
+                        type_override=InterpolationValidationError,
+                    )
+                return None
+            # Since we created a new node on the fly, future changes to this node are
+            # likely to be lost. We thus set the "readonly" flag to `True` to reduce
+            # the risk of accidental modifications.
+            wrapped._set_flag("readonly", True)
```

Comment:
I can see scenarios when this can get in the way.
As a future enhancement, we can consider supporting some flexibility around not setting this flag always.